File: klinara-ios/klinara-ios/Services/Notifications/MockMessagesService.swift
import Foundation

/// Sunucu olmadan mesaj günlüğünü sürmek için bellek-içi defter.
///
/// **Gerçek cursor sayfalaması yapar.** Tek sayfa döndüren bir mock, ekranın
/// sonsuz kaydırmasını ve ``MessageLogStore``ın "süzgeç değişince cursor'u
/// sıfırla" davranışını sınanmaz bırakırdı — ikisi de sessizce bozulabilecek
/// yerler.
final class MockMessagesService: MessagesService, @unchecked Sendable {

    /// Sunucudaki `DEFAULT_PAGE_SIZE` ile aynı değil: mock'ta sayfa küçük
    /// olmalı ki beş satırlık tohumda bile ikinci sayfa gerçekten oluşsun.
    private static let defaultPageSize = 3

    private let lock = NSLock()
    private var records: [Message] = []

    init() {
        seed()
    }

    func reseed() {
        withLock { seed() }
    }

    private func seed() {
        records = MockNotificationsSeed.messages(at: Date())
    }

    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    private func latency(_ seconds: Double = 0.3) async {
        try? await Task.sleep(for: .seconds(seconds))
    }

    /// Yeni bir mesaj üretildiğinde günlüğe düşsün diye — WhatsApp test
    /// gönderimi mock'u bunu çağırıyor. Gerçek sunucuda ikisi aynı
    /// `message_log` tablosuna yazıyor; mock'ta bağı kurmasak test gönderimi
    /// hiçbir yerde iz bırakmazdı.
    func record(_ message: Message) {
        withLock { records.insert(message, at: 0) }
    }

    func messages(
        cursor: String?,
        limit: Int?,
        filter: MessageFilter
    ) async throws -> Page<Message> {
        await latency()
        return withLock {
            // Sunucunun sırası: `created_at DESC, id DESC` — en yeni önce.
            let sorted = records
                .filter { matches($0, filter) }
                .sorted {
                    $0.createdAt == $1.createdAt ? $0.id > $1.id : $0.createdAt > $1.createdAt
                }
            let start = cursor.flatMap { key in
                sorted.firstIndex { "\($0.createdAt.timeIntervalSince1970)|\($0.id)" == key }
                    .map { $0 + 1 }
            } ?? 0
            let size = min(limit ?? Self.defaultPageSize, 200)
            let end = min(start + size, sorted.count)
            guard start < end else { return .empty }
            let page = Array(sorted[start..<end])
            let hasMore = end < sorted.count
            let next = hasMore
                ? page.last.map { "\($0.createdAt.timeIntervalSince1970)|\($0.id)" }
                : nil
            return Page(data: page, pageInfo: PageInfo(nextCursor: next, hasMore: hasMore))
        }
    }

    private func matches(_ message: Message, _ filter: MessageFilter) -> Bool {
        if let customerId = filter.customerId, message.customerId != customerId { return false }
        if let channel = filter.channel, message.channel != channel { return false }
        if let event = filter.event, message.event != event { return false }
        if let status = filter.status, message.status != status { return false }
        if let from = filter.from, let date = KlinaraCoding.parseTimestamp(from),
           message.createdAt < date {
            return false
        }
        if let to = filter.to, let date = KlinaraCoding.parseTimestamp(to),
           message.createdAt > date {
            return false
        }
        return true
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/notifications/MockMessagesService.kt
package com.klinara.android.services.notifications

import com.klinara.android.services.networking.ApiError
import com.klinara.android.services.networking.Page
import com.klinara.android.services.networking.PageInfo
import kotlinx.coroutines.delay
import java.time.Instant
import kotlin.random.Random

/**
 * Mesaj günlüğü mock'u — **gerçek keyset imleci** uygular (iOS `MockMessagesService` paritesi).
 *
 * Tek sayfa döndüren bir mock, ekranın sonsuz kaydırmasını ve "süzgeç değişince imleci
 * sıfırla" davranışını sınanmaz bırakırdı — ikisi de sessizce bozulabilecek yerler.
 */
class MockMessagesService(
    private val latencyEnabled: Boolean = true,
    private val random: Random = Random.Default,
    now: () -> Instant = Instant::now,
    /** Sunucunun 50'si DEĞİL: beş satırlık tohumda bile ikinci sayfa gerçekten oluşsun. */
    private val defaultPageSize: Int = DEFAULT_PAGE_SIZE,
) : MessagesService {
    var failing: Boolean = false

    /** Sayfa isteklerinden biri düşsün diye (testler: "sonraki sayfa hatası yutulmaz"). */
    var failNextPage: Boolean = false

    private val records: MutableList<Message> = MockNotificationsSeed.messages(now()).toMutableList()

    /**
     * Yeni bir mesaj günlüğe düşsün diye — WhatsApp test gönderimi mock'u bunu çağırıyor.
     * Gerçek sunucuda ikisi aynı `message_log` tablosuna yazıyor.
     */
    fun record(message: Message) {
        records.add(0, message)
    }

    override suspend fun messages(
        cursor: String?,
        limit: Int?,
        filter: MessageFilter,
    ): Page<Message> {
        if (latencyEnabled) delay(random.nextLong(MIN_LATENCY_MILLIS, MAX_LATENCY_MILLIS))
        if (failing) throw ApiError.Network()
        if (cursor != null && failNextPage) {
            failNextPage = false
            throw ApiError.Network()
        }

        // Sunucunun sırası: `created_at DESC, id DESC` — en yeni önce.
        val sorted =
            records
                .filter { it.matches(filter) }
                .sortedWith(compareByDescending<Message> { it.createdAt }.thenByDescending { it.id })
        val start = cursor?.let { key -> sorted.indexOfFirst { it.cursorKey() == key } + 1 } ?: 0
        val size = (limit ?: defaultPageSize).coerceAtMost(MAX_LIMIT)
        val end = (start + size).coerceAtMost(sorted.size)
        if (start >= end) return Page.empty()
        val page = sorted.subList(start, end)
        val hasMore = end < sorted.size
        return Page(page.toList(), PageInfo(nextCursor = page.last().cursorKey().takeIf { hasMore }, hasMore = hasMore))
    }

    private fun Message.cursorKey(): String = "${createdAt.toEpochMilli()}|$id"

    private fun Message.matches(filter: MessageFilter): Boolean =
        (filter.customerId == null || customerId == filter.customerId) &&
            (filter.channel == null || channel == filter.channel) &&
            (filter.event == null || event == filter.event) &&
            (filter.status == null || status == filter.status) &&
            (filter.from == null || createdAt >= filter.from) &&
            (filter.to == null || createdAt <= filter.to)

    private companion object {
        const val DEFAULT_PAGE_SIZE = 3
        const val MAX_LIMIT = 200
        const val MIN_LATENCY_MILLIS = 120L
        const val MAX_LATENCY_MILLIS = 400L
    }
}
